{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "script_folder='/home/dengw1/workspace/snakerun/MeCP2_iCLIP/scripts'\n",
    "data_folder='/home/dengw1/workspace/MeCP2_iCLIP'\n",
    "genome_folder='/home/dengw1/workspace/genome'\n",
    "sub_folder='{script_folder}/qsub'.format(script_folder=script_folder)\n",
    "if not os.path.isdir(sub_folder):\n",
    "    os.makedirs(sub_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Demultiplex\n",
    "dem='python {script_folder}/demultiplex_read1.py {data_folder}/raw_fq/Undetermined_S0_L003_R1_001.fastq {data_folder}/raw_fq'.format(script_folder=script_folder,data_folder=data_folder)\n",
    "print(subprocess.check_output(dem,shell=True))\n",
    "\n",
    "# Adapter triming, QC, STAR mapping & rRNA/tRNA masking\n",
    "cmd='''mkdir -p {data_folder}/cutadapt\n",
    "cutadapt -a AGATCGGAAGAGCGGTTCAG --quality-cutoff 6 -m 15 -o {data_folder}/cutadapt/{group}.fq {data_folder}/raw_fq/{group}.fq\n",
    "fastqc -o {data_folder}/cutadapt/ {data_folder}/cutadapt/{group}.fq\n",
    "mkdir -p {data_folder}/star/{group}/\n",
    "STAR --genomeDir {genome_folder}/star/{genome} \\\n",
    "--readFilesIn {data_folder}/cutadapt/{group}.fq  --outSAMtype BAM Unsorted \\\n",
    "--outFileNamePrefix {data_folder}/star/{group}/ \\\n",
    "--outFilterMultimapNmax 100 \\\n",
    "--runThreadN 20 \\\n",
    "--alignEndsProtrude 15 ConcordantPair \\\n",
    "--twopassMode Basic --limitOutSJcollapsed 2000000 \\\n",
    "--outStd Log > {data_folder}/star/{group}/log.txt 2>&1\n",
    "bedtools intersect -f 0.90 -abam {data_folder}/star/{group}/Aligned.out.bam -b {genome_folder}/raw/{genome}/rRNA_tRNA.bed -v > {data_folder}/star/{group}/Aligned.out.mask_rRNA.bam\n",
    "python {script_folder}/mapping_stat.py {data_folder}/star/{group}/Aligned.out.bam {data_folder}/cutadapt/{group}_fastqc.zip > {data_folder}/star/{group}/mapping_stat.txt\n",
    "python {script_folder}/collapse_dup.py -b {data_folder}/star/{group}/Aligned.out.mask_rRNA.bam -o {data_folder}/star/{group}/Aligned.out.mask_rRNA.collapsed_dup.bam -m {data_folder}/star/{group}/dup_removal.txt\n",
    "mkdir -p {data_folder}/clam/{group}\n",
    "CLAM preprocessor -i {data_folder}/star/{group}/Aligned.out.mask_rRNA.collapsed_dup.bam -o {data_folder}/clam/{group} --read-tagger-method start --lib-type sense\n",
    "'''\n",
    "for group in ['M1','M2','M3','H1','H2','H3']:\n",
    "    genome='hg19'\n",
    "    if group.startswith('M'):\n",
    "        genome='mm10'\n",
    "    cmd_f='%s/qsub/%s.sh'%(script_folder,group)\n",
    "    with open(cmd_f,'w') as cmd_file:\n",
    "        cmd_file.write(cmd.format(group=group,genome=genome,data_folder=data_folder,genome_folder=genome_folder,script_folder=script_folder))\n",
    "    print(subprocess.check_output('qsub -cwd -l h_vmem=40G,m_mem_free=40G %s'%cmd_f,shell=True))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd='''genome={genome}\n",
    "\n",
    "mkdir -p {data_folder}/clam/{group}_{binw}\n",
    "mkdir -p {data_folder}/clam/peaks_{binw}/{group}\n",
    "mkdir -p {data_folder}/homer/{binw}/{group}\n",
    "\n",
    "CLAM realigner -i {data_folder}/clam/{group}/multi.sorted.bam -o {data_folder}/clam/{group}_{binw} --winsize {binw} --max-tags -1 --read-tagger-method start --lib-type sense\n",
    "\n",
    "echo 'realigned'\n",
    "CLAM peakcaller -i {data_folder}/clam/{group}/unique.sorted.bam {data_folder}/clam/{group}_{binw}/realigned.sorted.bam \\\n",
    "-o {data_folder}/clam/peaks_{binw}/{group}  --binsize {binw} --qval-cutoff 1 \\\n",
    "--gtf {genome_folder}/raw/{genome}/{genome}.gtf -p 20\n",
    "\n",
    "python {script_folder}/make_bw.py --ub {data_folder}/clam/{group}/unique.sorted.bam --rb \\\n",
    "    {data_folder}/clam/{group}_{binw}/realigned.sorted.bam -g {genome_folder}/raw/hg19/hg19.gtf -u -t 20 -o {data_folder}/bigwig/{binw} -s {group} -c {genome_folder}/raw/{genome}/chr_size.txt\n",
    "\n",
    "CLAM permutation_callpeak -i {data_folder}/clam/{group}/unique.sorted.bam {data_folder}/clam/{group}_{binw}/realigned.sorted.bam \\\n",
    "-o {data_folder}/clam/peaks_{binw}/{group} -p 30 \\\n",
    "--gtf {genome_folder}/raw/{genome}/{genome}.gtf --qval-cutoff 1 --merge-size {binw}\n",
    "\n",
    "mv {data_folder}/clam/peaks_{binw}/{group}/narrow_peak.permutation.bed {data_folder}/clam/peaks_{binw}/{group}/narrow_peak.permutation.bed.all\n",
    "awk '{{split($4,a,\":\");if(a[3]<=0.005){{print}}}}' {data_folder}/clam/peaks_{binw}/{group}/narrow_peak.permutation.bed.all > {data_folder}/clam/peaks_{binw}/{group}/narrow_peak.permutation.bed\n",
    "\n",
    "echo 'peak_called'\n",
    "findMotifsGenome.pl {data_folder}/clam/peaks_{binw}/{group}/narrow_peak.permutation.bed {genome} {data_folder}/homer/{binw}/{group} \\\n",
    "-rna -len 5,6,7 \\\n",
    "-p 20 -size 100 -S 10\n",
    "echo 'done'\n",
    "'''\n",
    "for group in ['M1','M2','M3','H1','H2','H3']:\n",
    "    genome='hg19'\n",
    "    if group.startswith('M'):\n",
    "        genome='mm10'\n",
    "    for binw in ['100','150','200']:\n",
    "        cmd_f='%s/qsub/%s_%s_clam.sh'%(script_folder,group,binw)\n",
    "        with open(cmd_f,'w') as cmd_file:\n",
    "            cmd_file.write(cmd.format(group=group,genome=genome,binw=binw,data_folder=data_folder,genome_folder=genome_folder,script_folder=script_folder))\n",
    "        print(group+'_'+binw)\n",
    "        print(subprocess.check_output('qsub -cwd -l h_vmem=80G,m_mem_free=40G %s'%cmd_f,shell=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "summary=defaultdict(lambda:defaultdict(str))\n",
    "print(data_folder)\n",
    "for group in ['M1','M2','M3','H1','H2','H3']:\n",
    "    for line in open('%s/star/%s/Log.final.out'%(data_folder,group)):\n",
    "        info=line.strip().split('\\t')\n",
    "        if 'Number of input reads' in line:\n",
    "            summary[group]['input']=info[1]\n",
    "        elif 'Average mapped length' in line:\n",
    "            summary[group]['mapped_len']=info[1]\n",
    "        elif 'Uniquely mapped reads' in line:\n",
    "            summary[group]['unique_rate']=info[1]\n",
    "        elif r'% of reads mapped to multiple loci' in line:\n",
    "            summary[group]['multi_rate']=info[1]\n",
    "    summary[group]['masked']=subprocess.check_output('samtools view -c -F 260 %s/star/%s/Aligned.out.mask_rRNA.bam'%(data_folder,group),shell=True).decode('utf-8').strip()\n",
    "    summary[group]['deduped']=subprocess.check_output('samtools view -c -F 260 %s/star/%s/Aligned.out.mask_rRNA.collapsed_dup.bam'%(data_folder,group),shell=True).decode('utf-8').strip()\n",
    "    summary[group]['realigned']=subprocess.check_output('samtools view -c -F 260 %s/clam/%s_200/realigned.sorted.bam'%(data_folder,group),shell=True).decode('utf-8').strip()\n",
    "    summary[group]['peak_unique']=subprocess.check_output('samtools view -c -F 260 %s/clam/%s/unique.sorted.bam'%(data_folder,group),shell=True).decode('utf-8').strip()\n",
    "pd.DataFrame.from_dict(dict(summary)).to_csv('results/summary.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}